feat(scripts): add script to seed new address whitelist

chrismaree opened the pull request:
**Motivation**

When deploying to new chains it is painful to setup the new address whitelist. this PR helps this situation by providing a script to seed new address whitelists on new target chains.

The fetches the existing whitelist addresses on L1 (filtering out removed tokens) and then finds the associated L2 token addresses. for now the script works on arbitrum, optimism and boba.
for optimism and boba the scrip pulls addresses from the L2StandardBridge , looking for DepositFinalized events for the associated L1 token. On optimism and Arbitrum I also found these token lists ([arbitrum](https://bridge.arbitrum.io/token-list-42161.json) and [optimism](https://static.optimism.io/optimism.tokenlist.json)) from the [uniswap source code](https://github.com/Uniswap/interface/blob/main/src/constants/lists.ts) that includes some other tokens. using a combination of contract events and inclusion in these lists I think I’ve hit about 95% of all the tokens that are: a) in our whitelist and b) on the associated l2. you can see the outputs below:

Arbitrum:
![image](https://user-images.githubusercontent.com/12886084/148532519-9bfe7a4c-6dc2-4881-b593-67b7393af33d.png)

Optimism:
![image](https://user-images.githubusercontent.com/12886084/148532552-cbb9b7b3-6e46-46d5-b168-6ede31022dab.png)

Boba:
![image](https://user-images.githubusercontent.com/12886084/148532574-fd2f9f56-c106-43de-b21e-17375ee4eab0.png)


**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [X]  Untested

nicholaspai reviewed (commented):
High level comment: do you think this script is better placed in `scripts` or in the `common/hardhat/tasks` folder? The latter would be consistent with the `migrate-identifiers` script that I've used to do a similar logic in copying identifiers from one contract to another

nicholaspai commented on packages/scripts/src/multi-chain/BatchCollateralWhitelist.js:
```diff
@@ -0,0 +1,75 @@
+  const l2Web3 = getWeb3ByChainId(l2ChainId);
+  const l1Account = (await l1Web3.eth.getAccounts())[0];
+
+  console.log(`Running OVM Batch Collateral whitelister from ${l1ChainId}->${l2ChainId} with`);
```
I wouldn't name this just OVM batch whitelister, since it should work on other L2s right?

chrismaree replied:
ye, that's a mistake. when I started it was OVM spesific but I was able to generalize it.

nicholaspai commented on packages/scripts/src/multi-chain/utils.js:
```diff
@@ -0,0 +1,109 @@
+
+async function _findL2TokenFromTokenList(l2ChainId, l1TokenAddress) {
+  if (l2ChainId == 10) {
+    const response = await fetch("https://static.optimism.io/optimism.tokenlist.json");
```
if l2ChainId !== 10 or 42161 where will it error?

chrismaree replied:
if the chain id is neither of these then it returns a blank array from the token lists. for example boba's list is built up exclusively from the `_findL2TokenForOvmChain` method, not using any lists, as their is no token list like this for boba.

chrismaree commented:
> High level comment: do you think this script is better placed in `scripts` or in the `common/hardhat/tasks` folder? The latter would be consistent with the `migrate-identifiers` script that I've used to do a similar logic in copying identifiers from one contract to another

ye, I was not exactly sure where to put it but you are right that making this a hardhat task is better! I've refactored this to be a hard hat task, in ts, like the other ones and updated the documentation in `cross-chain-oracle` to reflect how to use this task.

nicholaspai commented on packages/common/src/hardhat/tasks/collateralWhitelist.ts:
```diff
@@ -0,0 +1,238 @@
+    const l1ChainId = Number(taskArguments.l1chainid);
+    const l2ChainId = Number(taskArguments.l2chainid);
+
+    const l1Web3 = new Web3(process.env[`NODE_URL_${l1ChainId}`] as string);
```
Since you're checking NODE_URL_CHAIN_ID you're already set up to use that well

chrismaree replied:
I originally did do this but I had issues with the web3 instance being returned. For some reason, the web3 instance from the `getWeb3(l2ChainId)` could not accept sending of transactions. I would get the following error: `The method eth_sendTransaction does not exist/is not available`. I've seen this before; it happens when you try send a raw transaction. for some or other reason the web3 instances instantiated like this was sending txs like this. 

The solution to this was to use `web3` from the `hre` and create a new web3 instance for l1web3. I could use `getWeb3(l1ChainId)` but it's functionally equivalent to what I've got here without needed an extra import. The way the code is here is more similar to the other hardhat tasks as well.

nicholaspai commented on packages/common/src/hardhat/tasks/collateralWhitelist.ts:
```diff
@@ -0,0 +1,238 @@
+    const l2TokenWhitelistContract = await deployments.get("AddressWhitelist");
+    const l2TokenWhitelist = new l2Web3.eth.Contract(l2TokenWhitelistContract.abi, l2TokenWhitelistContract.address);
+    const currentBlock = await l2Web3.eth.getBlockNumber();
+    const blockLookBack = 99999; // We need to use paginated query on L2 as some L2s limit how far you can look back
```
I like to use 99900 to be safer

nicholaspai commented on packages/common/src/hardhat/tasks/collateralWhitelist.ts:
```diff
@@ -0,0 +1,238 @@
+      const whitelistTx = await l2TokenWhitelist.methods
+        .addToWhitelist(filteredCombinedSet[index].l2TokenAddress)
+        .send({ from: deployer, nonce });
+      nonce++;
```
Why are you setting nonce? Can't you just use `.send({ from: })`?

chrismaree replied:
for example I will often get this:
![image](https://user-images.githubusercontent.com/12886084/148774965-8bc20acf-ab23-4020-b84c-ba6e43fd8f48.png)

nicholaspai reviewed (commented):
Thanks for making this a task!

mrice32 reviewed (commented):
Looks good! Just one comment on the env/getWeb3 usage.
